test/e2e/network/topology_hints.go

```diff
@@ -0,0 +1,188 @@
+			if listErr != nil {
+				return false, listErr
+			}
+
+			numEndpoints := 0
+			for _, slice := range slices.Items {
+				numEndpoints += len(slice.Endpoints)
+			}
+			if len(nodeNames) > numEndpoints {
+				framework.Logf("Expected %d endpoints, got %d", len(nodeNames), numEndpoints)
+				return false, nil
+			}
+
+			finalSlices = slices.Items
+			return true, nil
+		})
+		framework.ExpectNoError(err, "timed out waiting for EndpointSlices to be ready")
+
+		ginkgo.By("having hints set for each endpoint")
+		for _, slice := range finalSlices {
+			for _, ep := range slice.Endpoints {
+				if ep.Zone == nil {
+					framework.Failf("Expected endpoint in %s to have zone: %v", slice.Name, ep)
+				}
+				if ep.Hints == nil || len(ep.Hints.ForZones) == 0 {
+					framework.Failf("Expected endpoint in %s to have hints: %v", slice.Name, ep)
+				}
+				if len(ep.Hints.ForZones) > 1 {
+					framework.Failf("Expected endpoint in %s to have exactly 1 zone hint, got %d: %v", slice.Name, len(ep.Hints.ForZones), ep)
+				}
+				if *ep.Zone != ep.Hints.ForZones[0].Name {
+					framework.Failf("Expected endpoint in %s to have same zone hint, got %s: %v", slice.Name, *ep.Zone, ep)
+				}
+			}
+		}
+
+		nodeList, err := c.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+		framework.ExpectNoError(err)
+		nodesByZone := map[string]string{}
+		for _, node := range nodeList.Items {
+			if zone, ok := node.Labels[v1.LabelTopologyZone]; ok {
+				nodesByZone[node.Name] = zone
+			}
+		}
+
+		podList, err := c.CoreV1().Pods(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{})
+		framework.ExpectNoError(err)
+		podsByZone := map[string]string{}
+		for _, pod := range podList.Items {
+			if zone, ok := nodesByZone[pod.Spec.NodeName]; ok {
+				podsByZone[pod.Name] = zone
+			}
+		}
+
+		ginkgo.By("keeping requests in the same zone")
+		for i, nodeName := range nodeNames {
+			// Iterate through max of 3 nodes
+			if i > 2 {
+				break
+			}
+			fromZone, ok := nodesByZone[nodeName]
+			if !ok {
+				framework.Failf("Expected zone to be specified for %s node", nodeName)
+			}
+			execPod := e2epod.CreateExecPodOrFail(f.ClientSet, f.Namespace.Name, "execpod-", func(pod *v1.Pod) {
+				pod.Spec.NodeName = nodeName
+			})
+			framework.Logf("Ensuring that requests from %s pod on %s node stay in %s zone", execPod.Name, nodeName, fromZone)
+
+			cmd := fmt.Sprintf("curl -q -s --connect-timeout 2 http://%s:80/", svc.Name)
+			consecutiveRequests := 0
+			var stdout string
+			if pollErr := wait.PollImmediate(framework.Poll, e2eservice.KubeProxyLagTimeout, func() (bool, error) {
+				var err error
+				stdout, err = framework.RunHostCmd(f.Namespace.Name, execPod.Name, cmd)
+				if err != nil {
+					framework.Logf("unexpected error running %s from %s", cmd, execPod.Name)
+					return false, nil
+				}
```

Comment:
doing an "exec" per iteration is expensive and a source of flakiness,  it is more robust to build just one command that runs all the requests in a loop

https://github.com/kubernetes/kubernetes/blob/08bf54678e2bef8b4c72c8c6d104b0bdf19a6f75/test/e2e/network/service.go#L128-L138

I also will try to go for higher numbers of requests to be more resilient , 1 error from 5 is much problematic than 1 of 50 :smile:

Reply:
Chatted about this in Slack and clarified that it's not so much a percentage of successful requests as it is a number of consecutive successful requests that this is looking for. The whole idea here is that it could take some time for kube-proxy to correctly program iptables rules. Once it does, we want to ensure that all requests are routed within the same zone.

@aojea had the great idea to read pod logs instead of either a complicated shell script or multiple kubectl execs. That's the approach I've updated the PR to use. In my limited testing, the full e2e run for this test is ~35% faster with this approach, and also likely much less prone to flakiness.